generation/src/generation.rs

```diff
@@ -427,6 +483,104 @@ impl Register {
         }
         .into_token_stream()
     }
+
+    fn generate_block_definition(
+        &self,
+        name: &str,
+        description: Option<&str>,
+        registers: &RegisterCollection,
+        device: &Device,
+        device_impl: &syn::ItemImpl,
+    ) -> TokenStream {
+        let cfg_attributes = &self.cfg_attributes;
+
+        let pascal_case_name_string = name.to_case(Case::Pascal);
+        let pascal_case_name = syn::Ident::new(&pascal_case_name_string, Span::call_site());
+        let snake_case_name = syn::Ident::new(&name.to_case(Case::Snake), Span::call_site());
+
+        let module_doc_string = format!("Implementation of the [{pascal_case_name}]({snake_case_name}::{pascal_case_name}) register block");
+
+        let doc_attribute = if let Some(description) = description.as_ref() {
+            quote! { #[doc = #description] }
+        } else {
+            quote!()
+        };
+
+        let address_type = if let Some(register_address_type) = device.register_address_type {
+            register_address_type.into_type()
+        } else {
+            return syn::Error::new(Span::call_site(), "register_address_type is not specified")
+                .to_compile_error();
+        };
+
+        let mut generics = device_impl.generics.clone();
+        generics
+            .params
+            .insert(0, syn::GenericParam::Lifetime(syn::parse_quote!('block)));
+        generics.params.push(syn::GenericParam::Const(
+            syn::parse_quote!(const BASE_ADDR: #address_type),
+        ));
+
+        let device_type = device_impl.self_ty.clone();
+        let (impl_generics, ty_generics, where_clause) = generics.split_for_impl();
+        let fns = self.generate_block_functions(registers);
+        let regs = self.generate_block_register_definitions(registers, device, device_impl);
+
+        quote! {
+            #[doc = #module_doc_string]
+            pub mod #snake_case_name {
+                use super::*;
+
+                #doc_attribute
+                #(#cfg_attributes)*
+                pub struct #pascal_case_name #generics (pub(super) &'block mut #device_type);
+
+                impl #impl_generics ::device_driver::RegisterBlock for #pascal_case_name #ty_generics #where_clause {
+                    type Device = #device_type;
+                    const BASE_ADDR: #address_type = BASE_ADDR;
+
+                    fn dev(&mut self) -> &mut Self::Device {
+                        self.0
+                    }
+                }
```

Comment:
If we're generating code anyways, does it make sense to use the trait?
It adds to the public API while not strictly required

Reply:
I'm not sure it would be possible to generate the appropriate bounds to properly implement `RegisterDevice`/`AsyncRegisterDevice` without the trait. You need to only impl `RegisterDevice` if the device itself impls `RegisterDevice`, but I don't think you can express that bound without an associated type or generic parameter which wouldn't be present for a concrete generated struct.

Reply:
I think it's very possible

```rust
struct Block<'d, D> {
    device: &'d mut D,
}

impl<'d, D: RegisterDevice> RegisterDevice for Block<'d, D> {
    // ...
}

// Same with async device
```

I think I like this better. It allows keeping the public api of the crate a bit cleaner.

Reply:
Thinking about it, I feel like the base address of a block should be stored as a field in the struct. This would allow reuse of the type for repeated and copied blocks